File: mobile_apps/flutter_legacy/lib/features/dao/dao_service.dart
// DAO Service - Flutter Mobile
// Governance and DAO functionality

import 'dart:convert';
import 'package:http/http.dart' as http;

class DAOService {
  static const String API_BASE = 'http://localhost:8443/api/v1';
  String? _token;
  
  DAOService({String? token}) : _token = token;
  
  Map<String, String> get _headers => {
    'Content-Type': 'application/json',
    if (_token != null) 'Authorization': 'Bearer $_token',
  };
  
  // Get DAOs
  Future<List<DAO>> getDAOs() async {
    final response = await http.get(
      Uri.parse('$API_BASE/dao/list'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] as List).map((d) => DAO.fromJson(d)).toList();
    }
    return [];
  }
  
  // Get DAO proposals
  Future<List<Proposal>> getProposals(String daoId) async {
    final response = await http.get(
      Uri.parse('$API_BASE/dao/$daoId/proposals'),
      headers: _headers,
    );
    
    if (response.statusCode == 200) {
      final data = json.decode(response.body);
      return (data['data'] as List).map((p) => Proposal.fromJson(p)).toList();
    }
    return [];
  }
  
  // Create proposal
  Future<Proposal> createProposal(String daoId, String title, String description, String type) async {
    final response = await http.post(
      Uri.parse('$API_BASE/dao/$daoId/proposals'),
      headers: _headers,
      body: json.encode({
        'title': title,
        'description': description,
        'type': type,
      }),
    );
    
    if (response.statusCode == 201) {
      final data = json.decode(response.body);
      return Proposal.fromJson(data['data']);
    }
    throw Exception('Failed to create proposal');
  }
  
  // Vote
  Future<bool> vote(String proposalId, String vote, double weight) async {
    final response = await http.post(
      Uri.parse('$API_BASE/dao/proposals/$proposalId/vote'),
      headers: _headers,
      body: json.encode({'vote': vote, 'weight': weight}),
    );
    
    return response.statusCode == 200;
  }
  
  // Execute proposal
  Future<bool> executeProposal(String proposalId) async {
    final response = await http.post(
      Uri.parse('$API_BASE/dao/proposals/$proposalId/execute'),
      headers: _headers,
    );
    
    return response.statusCode == 200;
  }
  
  // Delegate voting power
  Future<bool> delegate(String daoId, String delegatee) async {
    final response = await http.post(
      Uri.parse('$API_BASE/dao/$daoId/delegate'),
      headers: _headers,
      body: json.encode({'delegatee': delegatee}),
    );
    
    return response.statusCode == 200;
  }
}

class DAO {
  final String id;
  final String name;
  final String description;
  final String token;
  final String treasuryAddress;
  final int memberCount;
  final double treasuryValue;
  
  DAO({
    required this.id,
    required this.name,
    required this.description,
    required this.token,
    required this.treasuryAddress,
    required this.memberCount,
    required this.treasuryValue,
  });
  
  factory DAO.fromJson(Map<String, dynamic> json) {
    return DAO(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      token: json['token'] ?? '',
      treasuryAddress: json['treasuryAddress'] ?? '',
      memberCount: json['memberCount'] ?? 0,
      treasuryValue: (json['treasuryValue'] ?? 0).toDouble(),
    );
  }
}

class Proposal {
  final String id;
  final String daoId;
  final String title;
  final String description;
  final String type;
  final String status;
  final double forVotes;
  final double againstVotes;
  final int quorum;
  final DateTime createdAt;
  final DateTime? executedAt;
  
  Proposal({
    required this.id,
    required this.daoId,
    required this.title,
    required this.description,
    required this.type,
    required this.status,
    required this.forVotes,
    required this.againstVotes,
    required this.quorum,
    required this.createdAt,
    this.executedAt,
  });
  
  factory Proposal.fromJson(Map<String, dynamic> json) {
    return Proposal(
      id: json['id'] ?? '',
      daoId: json['daoId'] ?? '',
      title: json['title'] ?? '',
      description: json['description'] ?? '',
      type: json['type'] ?? '',
      status: json['status'] ?? 'PENDING',
      forVotes: (json['forVotes'] ?? 0).toDouble(),
      againstVotes: (json['againstVotes'] ?? 0).toDouble(),
      quorum: json['quorum'] ?? 0,
      createdAt: DateTime.parse(json['createdAt']),
      executedAt: json['executedAt'] != null ? DateTime.parse(json['executedAt']) : null,
    );
  }
}

File: mobile_apps/android_legacy/app/src/main/java/com/tigerwallet/app/services/dao/DAOService.java
package com.tigerwallet.app.services;

import org.json.JSONArray;
import org.json.JSONObject;
import java.net.HttpURLConnection;
import java.net.URL;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

public class DAOService {
    private static final String API_BASE = "https://api.tigerwallet.com/api/v1";
    private String authToken;
    
    public DAOService(String token) { this.authToken = token; }
    
    public List<DAO> getDAOs() throws Exception {
        List<DAO> daos = new ArrayList<>();
        URL url = new URL(API_BASE + "/dao/list");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        conn.setRequestProperty("Content-Type", "application/json");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        if (conn.getResponseCode() == 200) {
            String response = readResponse(conn);
            JSONArray data = new JSONObject(response).getJSONArray("data");
            for (int i = 0; i < data.length(); i++) {
                JSONObject d = data.getJSONObject(i);
                daos.add(new DAO(d.getString("id"), d.getString("name"), d.getString("description"), d.getInt("memberCount")));
            }
        }
        return daos;
    }
    
    public List<Proposal> getProposals(String daoId) throws Exception {
        List<Proposal> proposals = new ArrayList<>();
        URL url = new URL(API_BASE + "/dao/" + daoId + "/proposals");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("GET");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        if (conn.getResponseCode() == 200) {
            String response = readResponse(conn);
            JSONArray data = new JSONObject(response).getJSONArray("data");
            for (int i = 0; i < data.length(); i++) {
                JSONObject p = data.getJSONObject(i);
                proposals.add(new Proposal(p.getString("id"), p.getString("title"), p.getString("description"), p.getString("status"), p.getDouble("forVotes"), p.getDouble("againstVotes")));
            }
        }
        return proposals;
    }
    
    public boolean vote(String proposalId, String choice, double weight) throws Exception {
        URL url = new URL(API_BASE + "/dao/proposals/" + proposalId + "/vote");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        conn.setDoOutput(true);
        JSONObject body = new JSONObject();
        body.put("choice", choice);
        body.put("weight", weight);
        conn.getOutputStream().write(body.toString().getBytes(StandardCharsets.UTF_8));
        return conn.getResponseCode() == 200;
    }
    
    public boolean createProposal(String daoId, String title, String description, String type) throws Exception {
        URL url = new URL(API_BASE + "/dao/" + daoId + "/proposals");
        HttpURLConnection conn = (HttpURLConnection) url.openConnection();
        conn.setRequestMethod("POST");
        conn.setRequestProperty("Content-Type", "application/json");
        if (authToken != null) conn.setRequestProperty("Authorization", "Bearer " + authToken);
        conn.setDoOutput(true);
        JSONObject body = new JSONObject();
        body.put("title", title);
        body.put("description", description);
        body.put("type", type);
        conn.getOutputStream().write(body.toString().getBytes(StandardCharsets.UTF_8));
        return conn.getResponseCode() == 201;
    }
    
    private String readResponse(HttpURLConnection conn) throws Exception {
        BufferedReader reader = new BufferedReader(new InputStreamReader(conn.getInputStream(), StandardCharsets.UTF_8));
        StringBuilder response = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) response.append(line);
        reader.close();
        return response.toString();
    }
    
    public static class DAO { public String id, name, description; public int memberCount; public DAO(String i, String n, String d, int m) { id=i; name=n; description=d; memberCount=m; } }
    public static class Proposal { public String id, title, description, status; public double forVotes, againstVotes; public Proposal(String i, String t, String d, String s, double f, double a) { id=i; title=t; description=d; status=s; forVotes=f; againstVotes=a; } }
}
